crates/nu-protocol/src/value/into_value.rs

```diff
@@ -0,0 +1,206 @@
+/// [`Value::Record`], where each field of the record corresponds to a field of the struct.
+/// For structs with unnamed fields, the value representation will be [`Value::List`], with all
+/// fields inserted into a list.
+/// Unit structs will be represented as [`Value::Nothing`] since they contain no data.
+///
+/// Only enums with no fields may derive this trait.
+/// The resulting value representation will be the name of the variant as a [`Value::String`].
+/// By default, variant names will be converted to ["snake_case"](convert_case::Case::Snake).
+/// You can customize the case conversion using `#[nu_value(rename_all = "kebab-case")]` on the enum.
+/// All deterministic case conversions provided by [`convert_case::Case`] are supported by
+/// specifying the case name followed by "case".
+///
+/// ```
+/// # use nu_protocol::{IntoValue, Value};
+/// #[derive(IntoValue)]
+/// #[nu_value(rename_all = "COBOL-CASE")]
+/// enum Bird {
+///     MountainEagle,
+///     ForestOwl,
+///     RiverDuck,
+/// }
+///
+/// assert_eq!(
+///     Bird::RiverDuck.into_value_unknown(),
+///     Value::test_string("RIVER-DUCK")
+/// );
+/// ```
+pub trait IntoValue: Sized {
+    /// Converts the given value to a [`Value`].
+    fn into_value(self, span: Span) -> Value;
+
+    /// Converts the given value to a [`Value`] with an unknown [`Span`].
+    ///
+    /// Internally this simply calls [`Span::unknown`] for the `span`.
+    fn into_value_unknown(self) -> Value {
+        Self::into_value(self, Span::unknown())
+    }
+}
+
+// Primitive Types
+
+impl<T, const N: usize> IntoValue for [T; N]
+where
+    T: IntoValue,
+{
+    fn into_value(self, span: Span) -> Value {
+        Vec::from(self).into_value(span)
+    }
+}
+
+macro_rules! primitive_into_value {
+    ($type:ty, $method:ident) => {
+        impl IntoValue for $type {
+            fn into_value(self, span: Span) -> Value {
+                Value::$method(self, span)
+            }
+        }
+    };
+
+    ($type:ty as $as_type:ty, $method:ident) => {
+        impl IntoValue for $type {
+            fn into_value(self, span: Span) -> Value {
+                Value::$method(self as $as_type, span)
+            }
+        }
+    };
+}
+
+primitive_into_value!(bool, bool);
+primitive_into_value!(char, string);
+primitive_into_value!(f32 as f64, float);
+primitive_into_value!(f64, float);
+primitive_into_value!(i8 as i64, int);
+primitive_into_value!(i16 as i64, int);
+primitive_into_value!(i32 as i64, int);
+primitive_into_value!(i64, int);
+primitive_into_value!(isize as i64, int);
+primitive_into_value!(u8 as i64, int);
+primitive_into_value!(u16 as i64, int);
+primitive_into_value!(u32 as i64, int);
+// u64 and usize may be truncated as Value only supports i64.
+
+impl IntoValue for () {
+    fn into_value(self, span: Span) -> Value {
+        Value::nothing(span)
+    }
+}
+
+macro_rules! tuple_into_value {
```

Comment:
I see your point but that is the same concern for all tuples in Rust. I think users should be allowed to decide for themselves. When `nu_protocol` doesn't implement them, no one will be able to use tuples because of the orphan rule.

Reply:
Fair enough, I guess we can keep the tuple implementations in case someone else needs them.